internal/model/db.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 The golang.design Initiative Authors.
+// All rights reserved. Use of this source code is governed
+// by a MIT license that can be found in the LICENSE file.
+//
+// Originally written by Changkun Ou <changkun.de> at
+// changkun.de/s/redir, adopted by Mai Yang <maiyang.me>.
+
+package model
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+var (
+	ErrExistedAlias = errors.New("alias is existed")
+)
+
+type AliasKind int
+
+const (
+	KindShort AliasKind = iota
+	KindRandom
+)
+
+const (
+	dbname   = "redir"
+	collink  = "links"
+	colvisit = "Visit"
+)
+
+// Redirect records a kind of alias and its correlated link.
+type Redirect struct {
+	Alias   string    `json:"alias"   bson:"alias"`
+	Kind    AliasKind `json:"kind"    bson:"kind"`
+	URL     string    `json:"url"     bson:"url"`
+	Private bool      `json:"private" bson:"private"`
+}
+
+// Visit indicates an Record of Visit pattern.
+type Visit struct {
+	Alias   string    `json:"alias"   bson:"alias"`
+	Kind    AliasKind `json:"kind"    bson:"kind"`
+	IP      string    `json:"ip"      bson:"ip"`
+	UA      string    `json:"ua"      bson:"ua"`
+	Referer string    `json:"referer" bson:"referer"`
+	Time    time.Time `json:"time"    bson:"time"`
+}
+
+type Refstat struct {
+	Referer string `json:"referer" bson:"referer"`
+	Count   int64  `json:"count"   bson:"count"`
+}
+
+type Uastat struct {
+	UA    string `json:"ua"    bson:"ua"`
+	Count int64  `json:"count" bson:"count"`
+}
+
+type Locstat struct {
+	Locations []string `bson:"locs" json:"locs"`
+}
+
+type Timehist struct {
+	Time  time.Time `bson:"time"  json:"time"`
+	Count int       `bson:"count" json:"count"`
+}
+
+type Record struct {
+	Alias string `bson:"alias"`
+	UV    int64  `bson:"uv"`
+	PV    int64  `bson:"pv"`
+}
+
+type RedirModel interface {
+	NewDB(url string) (RedirModel, error)
+	Close() (err error)
+}
+
+type RedirAliasDataModel interface {
+	StoreAlias(ctx context.Context, r *Redirect) (err error)
+	UpdateAlias(ctx context.Context, a, l string) (*Redirect, error)
+	DeleteAlias(ctx context.Context, a string) (err error)
+	FetchAlias(ctx context.Context, a string) (*Redirect, error)
```

Comment:
```suggestion
	StoreAlias(context.Context, r *Redirect) error
	UpdateAlias(context.Context, alias, link string) (*Redirect, error)
	DeleteAlias(context.Context, alias string) error
	FetchAlias(context.Context, alias string) (*Redirect, error)
```

在语意明确的情况下不写形参名，否则其他名字不用缩写，用于增加阅读的清晰度

Reply:
changkun/redir 的 alias 数据模型其实是有一定缺陷的，这个其实是继承自 golang-design/redir 增加了 random alias 的需求。具体来说 Alias+Kind 才能唯一确定一个 Alias ，然后才是更新、删除等操作。但是这里的接口没有考虑 Kind，相当于默认了 random alias 不会与带语义的 alias 重名。

可能有一个更好的设计会更好（也可能目前的情况也够用）